gobblin-api/src/main/java/gobblin/commit/CommitSequence.java

```diff
@@ -0,0 +1,153 @@
+ *
+ *    <pre>
+ *      {@code
+ *        CommitSequence sequence = CommitSequence.newBuilder()
+ *                                  .withJobName(jobName)
+ *                                  .withDatasetUrn(datasetUrn)
+ *
+ *                                  .buildStep(FsRenameCommitStep.Builder.class)
+ *                                  .withProps(props)
+ *                                  .from(srcPath)
+ *                                  .to(dstPath)
+ *                                  .stepDone()
+ *
+ *                                  .buildStep(DatasetStateCommitStep.Builder.class)
+ *                                  .withProps(props)
+ *                                  .withDatasetUrn(datasetUrn)
+ *                                  .withDatasetState(datasetState)
+ *                                  .stepDone()
+ *
+ *                                  .build();
+ *      }
+ *    </pre>
+ * </p>
+ *
+ * @author ziliu
+ */
+public class CommitSequence {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CommitSequence.class);
+
+  @Getter
+  private final String jobName;
+
+  @Getter
+  private final String datasetUrn;
+
+  private final List<CommitStep> steps;
+
+  private CommitSequence(Builder builder) {
+    this.jobName = builder.jobName;
+    this.datasetUrn = builder.datasetUrn;
+    this.steps = builder.steps;
+  }
+
+  public static class Builder {
+
+    private String jobName;
+    private String datasetUrn;
+    private final List<CommitStep> steps = Lists.newArrayList();
+
+    public Builder withJobName(String jobName) {
+      this.jobName = jobName;
+      return this;
+    }
+
+    public Builder withDatasetUrn(String datasetUrn) {
+      this.datasetUrn = datasetUrn;
+      return this;
+    }
+
+    /**
+     * Build a {@link CommitStep}.
+     *
+     * @param builderClass The builder class for the {@link CommitStep}, which should extend
+     * {@link CommitStepBase.Builder}.
+     * @return An instance of the builder class for the {@link CommitStep}.
+     */
+    public <T extends CommitStepBase.Builder<?>> T buildStep(Class<T> builderClass) {
+      try {
+        return ConstructorUtils.invokeConstructor(builderClass, this);
+      } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+        throw new RuntimeException("Failed to instantiate " + builderClass, e);
+      }
+    }
+
+    /**
+     * Add a {@link CommitStep} to the commit sequence.
+     */
+    public void addStep(CommitStep step) {
+      this.steps.add(step);
+    }
+
+    public CommitSequence build() {
+      return new CommitSequence(this);
+    }
+
+  }
+
+  /**
+   * Execute the {@link CommitStep}s in the order they are added to the commit sequence.
+   */
+  public void execute() {
+    try {
+      for (CommitStep step : this.steps) {
+        if (!step.isCompleted()) {
+          step.execute();
+        }
+      }
+    } catch (Throwable t) {
```

Comment:
Since the `CommitStep`s need to be executed in a particular order, once a step fails, no further steps should be executed.